### PR TITLE
@types/node: fix dns.lookup() definitions

### PR DESCRIPTION
**Pull request by ikokostya**
This is rebased pull request https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15103

### List of changes

* Fix `dns.lookup()` return type: instead of `string` return `void`. 
* Add signatures where `options` is object. There are 3 cases:
  1. Option `all` doesn't exist or `false`, then callback is `(err: NodeJS.ErrnoException, address: string, family: number) => void`.
  2. Option `all` is `true`, then callback is `(err: NodeJS.ErrnoException, addresses: Address[]) => void`.
  3. Options `all` has `boolean` type (for example, user assigns to option result of some function where we don't known exact literal type (`true` or `false`) at compile time), then callback is union of 1 and 2.
* Change type of `error` callback argument from `Error` to `NodeJS.ErrnoException`: https://github.com/nodejs/node/blob/98e54b0bd4854bdb3e2949d1b6b20d6777fb7cde/lib/dns.js#L184
* Export [getaddrinfo flags](https://nodejs.org/dist/latest-v6.x/docs/api/dns.html#dns_supported_getaddrinfo_flags).
* Add tests for this cases.
* Apply changes to the following `Node.js` versions: 7, 6, 4.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/dist/latest-v6.x/docs/api/dns.html#dns_dns_lookup_hostname_options_callback
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dslint/dt.json" }`.

### CODE REVIEW

**Comment by Flarna** [comment]
Looks like previous PR with requested changes included so fine from my point of view.

**Comment by hinell** [comment]
Same here. 
Previously I commented on ``lookup()``'s return type standing for not using anything there (i.e. to use ``void``) but from now on I think it would be better to use some return type like one the previous revision of this PR had for sake of [lookup node API](https://github.com/nodejs/node/blob/master/lib/dns.js#L176) compliance. 
Just my opinion. Anyway it is up to you to decide what it would return.
Right now it LGTM and I think it is ready to be merged.

File: types/node/index.d.ts [hunk 1]
```diff
@@ -1889,8 +1889,35 @@ declare module "dns" {
...
+
+    export interface LookupAddress {
+        address: string;
+        family: number;
+    }
+
+    export function lookup(domain: string, family: number, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupOneOptions, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupAllOptions, callback: (err: NodeJS.ErrnoException, addresses: LookupAddress[]) => void): void;
+    export function lookup(domain: string, options: LookupOptions, callback: (err: NodeJS.ErrnoException, address: string | LookupAddress[], family?: number) => void): void;
```
Comment (hinell) [inline types/node/index.d.ts]:
```ts
callback: (err: NodeJS.ErrnoException, address: string | LookupAddress[], family?: number) => void
```
Optional arguments  in the callback are senseless. Fix it:
```ts
 callback: (... family: number) ....
```

File: types/node/v4/index.d.ts [hunk 2]
```diff
@@ -1240,8 +1240,35 @@ declare module "url" {
...
+
+    export interface LookupAddress {
+        address: string;
+        family: number;
+    }
+
+    export function lookup(domain: string, family: number, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupOneOptions, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupAllOptions, callback: (err: NodeJS.ErrnoException, addresses: LookupAddress[]) => void): void;
+    export function lookup(domain: string, options: LookupOptions, callback: (err: NodeJS.ErrnoException, address: string | LookupAddress[], family?: number) => void): void;
```
Comment (hinell) [inline types/node/v4/index.d.ts]: Same

File: types/node/v6/index.d.ts [hunk 3]
```diff
@@ -1801,8 +1801,35 @@ declare module "dns" {
...
+
+    export interface LookupAddress {
+        address: string;
+        family: number;
+    }
+
+    export function lookup(domain: string, family: number, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupOneOptions, callback: (err: NodeJS.ErrnoException, address: string, family: number) => void): void;
+    export function lookup(domain: string, options: LookupAllOptions, callback: (err: NodeJS.ErrnoException, addresses: LookupAddress[]) => void): void;
+    export function lookup(domain: string, options: LookupOptions, callback: (err: NodeJS.ErrnoException, address: string | LookupAddress[], family?: number) => void): void;
```
Comment (hinell) [inline types/node/v6/index.d.ts]: And again

**Review by hinell — CHANGES_REQUESTED** [review]
Fix these and I'll approve.

**Review by hinell — APPROVED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (Flarna) [inline types/node/index.d.ts]: I don't agree here. Actually err can be null, address can be null or undefined and family can be undefined.
But it seems that this pattern is commonly used at a lot places in nodejs typings for some reason so maybe it's better to keep it in favour of a common style.

Comment (hinell) [inline types/node/index.d.ts]: It is a classical mistake. [Read more about here](http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks)

Comment (Flarna) [inline types/node/index.d.ts]: I have seen this before and to my understanding this tells that callback parameters should be marked optional if the callback may be called with partial parameters - like here.

Comment (hinell) [inline types/node/index.d.ts]: @Flarna Just try to read that article.

**Pull request merged**